doc/design_decisions.rst

```diff
@@ -0,0 +1,93 @@
+--------------------------------
+
+System state and history database hold similar data, but they are not interchangeable.
+That's why we must use them for the right purposes.
+
+* System state
+
+  * Holds information about the currently installed software (extends rpmdb and any other state information).
+  * Is used to retrieve information about installed software such as
+    ``reason``, ``from_repo``, list of installed comps groups and their packages etc.
+  * Consistency and reliability of the stored data is crucial for managing a working system.
+
+* History database
+
+  * Holds recorded transactions.
+  * Is used in ``history`` operations to rollback/undo/redo transactions.
+  * May be used to recover system state if it is lost.
+    System state may not be fully recovered in such case because of missing records in the history database.
+  * Important for auditing the transaction history.
+  * It is not crucial for managing the installed system and may be freely deleted (if auditing is not an issue).
+
+
+Handling installations made outside libdnf
+------------------------------------------
+
+Sometimes a user installs packages by directly using ``rpm`` command
+or an application that is not built on top of libdnf.
+By doing that, libdnf's system state data gets out of sync with installed packages.
+
+Consider installing an application and its dependencies using the ``rpm`` command.
+In DNF 4, all such packages would have ``unknown`` reason, which is equal to ``user``.
+This effectively disables removing the packages as unused dependencies
+and we need to explicitly provide all their names to the ``remove`` command.
+While this is quite easy for a power user, a normal user doesn't know if it's safe to
+remove the packages (especially low-level libraries)
+and ends up with a ever-growing package set they frequently solve by reinstalling the system from scratch.
+
+That's why we decided to create a differential transaction and update the system state and history database
+any time unexpected installed packages are detected. We detect leaf packages and mark them with reason ``user``,
+while keeping the remaining packages with reason ``dependency``.
+Removing the leaf package also removes its dependencies as expected.
+
+Unprivileged users can't change the system state and history database.
+We have decided to compute and keep the differential transaction in memory for them.
+When a privileged user runs a transaction, the differential transaction is stored on disk.
+
+
+Resolving reasons
+-----------------
+
+System state tracks the following:
+
+  * List of userinstalled packages.
+  * List of installed comps groups incl. packages that were installed with them.
+  * List of installed comps environments incl. groups that were installed with them.
+
+Package reason is resolved the following way:
+
+  * If a package is userinstalled, return reason ``user``.
+  * If a package is part of an installed group, return reason ``group``.
```

Comment:
These two sentences refer to two different sets of packages:
- List of installed comps groups incl. packages that were installed with them.
- If a package is part of an installed group, return reason ``group``.

What about packages that were installed with a group, but are no longer part of the group (because it was upgraded)? System state tracks them, but they have reason ``dependency``? What does it then mean that system state tracks them?
What about packages that are part of a group, but were not installed with it, but later as dependencies? Should they have reason ``group`` when user excluded them during the group installation?